### ISSUE TITLE
ArcFace easy margin

### ISSUE THREAD

**Issue by kevinlu1211** [ronghuaiyang/arcface-pytorch#24]
I was wondering in line 43 of [https://github.com/ronghuaiyang/arcface-pytorch/blob/master/models/metrics.py](metric.py) 

```
phi = torch.where(cosine > self.th, phi, cosine - self.mm)
```

`let th = cos(pi - margin)`
`let mm = margin * sin(pi -  margin)`

If it isn't too much trouble, could you explain why you do `cosine - mm` if `phi` is lower than `cos(pi - margin)`?  I understand if `cosine < th` then `phi < cosine` rendering the margin counterproductive, but is there a special geometric meaning by decreasing the value of `cosine` by `mm` or was this an arbitrary choice? 

I've looked at the ArcFace paper but I didn't see any mention of an easy margin, do you have any references for this?
Also, if I am not mistaken I think that the `ArcFaceLoss` only uses:

```
loss1 = self.classify_loss(output, labels)
```
and not 

```
loss2 = self.classify_loss(cosine, labels)
```
which I assume is the standard softmax loss. Was this just to help with the training process?

**Comment by shirley-wu** [comment]
I'm also confused so I checked the existing issues. I think issue #2 solved my problem and I hope it can help you.

**Comment by yuyijie1995** [comment]
@kevinlu1211 Have you figure out this problem? I am still confused when I read the #2

**Comment by kevinlu1211** [comment]
No I haven’t I’m also still confused haha

**Comment by fyang93** [comment]
It confused me too. What's the exact mathematical meaning of `self.mm`?

**Comment by ronghuaiyang** [comment]
the self.mm is like m in the cosface. when theta + m >=180, we use cosface instead.

**Comment by lizhenstat** [comment]
hi, [this issue](https://github.com/deepinsight/insightface/issues/108) gives out a more detailed explaination.
The plot can be drawn through the following code
```pyth# insightface-issue-108: draw pictures
import math
from math import cos, sin, pi
import numpy as np

m = 0.5
th = math.cos(math.pi - m)
mm = math.sin(math.pi - m) * m
cos_m = math.cos(m)
sin_m = math.sin(m)

tmp = np.linspace(0,pi,100)
cosine = [cos(c) for c in tmp]

y = []
for c in cosine:
    sine = math.sqrt((1.0 - c**2)) # .clamp(0, 1))
    phi = c * cos_m - sine * sin_m
    if c > th:
       y = y+[phi]
    else:
       y = y+[c - mm]


import matplotlib.pyplot as plt 
plt.plot(tmp,y)
plt.show()on

```

**Comment by twotwoiscute** [comment]
> hi, [this issue](https://github.com/deepinsight/insightface/issues/108) gives out a more detailed explaination.
> The plot can be drawn through the following code
> 
> ```
> import math
> from math import cos, sin, pi
> import numpy as np
> 
> m = 0.5
> th = math.cos(math.pi - m)
> mm = math.sin(math.pi - m) * m
> cos_m = math.cos(m)
> sin_m = math.sin(m)
> 
> tmp = np.linspace(0,pi,100)
> cosine = [cos(c) for c in tmp]
> 
> y = []
> for c in cosine:
>     sine = math.sqrt((1.0 - c**2)) # .clamp(0, 1))
>     phi = c * cos_m - sine * sin_m
>     if c > th:
>        y = y+[phi]
>     else:
>        y = y+[c - mm]
> 
> 
> import matplotlib.pyplot as plt 
> plt.plot(tmp,y)
> plt.show()on
> ```

Do you know the purpose of "making cos(theta+m) monotonic decreasing" ? Thanks !

**Comment by lizhenstat** [comment]
@twotwoiscute hi, the Section 3.2 in  [Sphereface](https://arxiv.org/pdf/1704.08063.pdf) gives why **phi**  needs to drease, I think it is the same reason here.

**Comment by twotwoiscute** [comment]
> @twotwoiscute hi, the Section 3.2 in [Sphereface](https://arxiv.org/pdf/1704.08063.pdf) gives why **phi** needs to drease, I think it is the same reason here.

Wow, it's really helpful! Thanks!
